spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/procedures/CreateChangeViewProcedure.java

```diff
@@ -0,0 +1,267 @@
+ *   <li>(id=1, data='b', op='INSERT')
+ * </ul>
+ *
+ * <p>will be marked as update-rows:
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='UPDATE_BEFORE')
+ *   <li>(id=1, data='b', op='UPDATE_AFTER')
+ * </ul>
+ */
+public class CreateChangeViewProcedure extends BaseProcedure {
+  private static final Logger LOG = LoggerFactory.getLogger(CreateChangeViewProcedure.class);
+
+  private static final ProcedureParameter[] PARAMETERS =
+      new ProcedureParameter[] {
+        ProcedureParameter.required("table", DataTypes.StringType),
+        ProcedureParameter.optional("changelog_view", DataTypes.StringType),
+        ProcedureParameter.optional("options", STRING_MAP),
+        ProcedureParameter.optional("compute_updates", DataTypes.BooleanType),
+        ProcedureParameter.optional("remove_carryovers", DataTypes.BooleanType),
+        ProcedureParameter.optional("identifier_columns", DataTypes.StringType),
+      };
+
+  private static final StructType OUTPUT_TYPE =
+      new StructType(
+          new StructField[] {
+            new StructField("changelog_view", DataTypes.StringType, false, Metadata.empty())
+          });
+
+  public static SparkProcedures.ProcedureBuilder builder() {
+    return new BaseProcedure.Builder<CreateChangeViewProcedure>() {
+      @Override
+      protected CreateChangeViewProcedure doBuild() {
+        return new CreateChangeViewProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private CreateChangeViewProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    String tableName = args.getString(0);
+
+    // Read data from the table.changes
+    Dataset<Row> df = changelogRecords(tableName, readOptions(args));
+
+    // compute remove carry-over rows by default
+    boolean removeCarryoverRow = args.isNullAt(4) ? true : args.getBoolean(4);
+
+    if (computeUpdatedRow(args)) {
+      String[] identifierColumns = identifierColumns(args, tableName);
+
+      Preconditions.checkArgument(
+          identifierColumns.length > 0,
+          "Cannot compute the update-rows because identifier columns are not set");
+
+      Column[] repartitionColumns = getRepartitionExpr(df, identifierColumns);
+      df = transform(df, repartitionColumns);
+    } else if (removeCarryoverRow) {
+      df = removeCarryoverRows(df);
+    }
+
+    String viewName = viewName(args, tableName);
+
+    // Create a view for users to query
+    df.createOrReplaceTempView(viewName);
+
+    return toOutputRows(viewName);
+  }
+
+  private boolean computeUpdatedRow(InternalRow args) {
+    if (!args.isNullAt(5)) {
+      return true;
+    }
+
+    return args.isNullAt(3) ? false : args.getBoolean(3);
+  }
+
+  private Dataset<Row> removeCarryoverRows(Dataset<Row> df) {
```

Comment:
Can you explore the idea of using native `dropDuplicates` operation instead of applying the iterator? At least, this should avoid the conversion from the internal to the public row representation, which is required by the iterator at the moment.

Reply:
Tested with `dropDuplicates(columns)`.  It doesn't fit here due to following reasons
1. It keeps one duplications. for example, we want to remove both rows, (1, 'data', 'DETELE') (1, 'data', 'INSERT'), but it keeps one.
2. In case of duplicated rows in the iceberg table, we still want to keep them. for example, we got 4 rows: (1, 'data', 'DETELE') (1, 'data', 'INSERT'), (1, 'data', 'INSERT'), (1, 'data', 'INSERT'). We want to only remove the first two rows. The `dropDuplicates` will remove all except keeping one. This is the behavior we don't want.

The iterator meet our requirements here. I will explore the window function. It should be faster.

Reply:
You are right, let's consider the best way for removing carryovers separately. Out of scope for this PR.